app/components/edit/EditNotes.tsx

```diff
@@ -12,139 +12,94 @@ export interface InternalNoteChanges {
+  const internalNotes = notes.map((note) => {
+    const newNote: InternalNote = {
+      ...note,
+      key: note.id,
     };
 
-    this.renderNotes = this.renderNotes.bind(this);
-    this.handleNoteChange = this.handleNoteChange.bind(this);
-    this.addNote = this.addNote.bind(this);
-    this.removeNote = this.removeNote.bind(this);
-  }
-
-  handleNoteChange(key: number, note: InternalNoteChanges): void {
-    const { notes } = this.state;
-    const { handleNotesChange } = this.props;
-    notes[key] = note;
-    this.setState(
-      {
-        notes,
-      },
-      () => {
-        // HACK: This is really strange, to pass in the entirety of the State of
-        // this component to a change handler. We should really be more
-        // selectively deciding what state should be synchronized to the parent
-        // component.
-        handleNotesChange(this.state);
-      }
-    );
-  }
-
-  addNote() {
-    const { existingNotes, uuid } = this.state;
-    const newUUID = uuid - 1;
-    existingNotes.push({
-      key: newUUID,
+    return newNote;
+  });
+
+  const [notesList, setNotesList] = useState(internalNotes);
+  const [uuid, setUuid] = useState(-1);
+  const [notesObject, setNotesObject] = useState<NotesObject>({});
+
+  useEffect(() => {
+    handleNotesChange(notesObject);
+  }, [notesObject, handleNotesChange]);
+
+  const handleNoteChange = (key: number, note: InternalNoteChanges): void => {
+    setNotesObject({
+      ...notesObject,
+      [key]: note,
     });
-    this.setState({ existingNotes, uuid: newUUID });
-  }
-
-  removeNote(index: number) {
-    const { handleNotesChange } = this.props;
-    const { existingNotes, notes } = this.state;
-    const currentNote = existingNotes[index];
-    currentNote.isRemoved = true;
-    const { key } = currentNote;
-    // HACK: This is really bad, and we shouldn't be directly mutating the state
-    // like this. Instead we should create a copy of the state variables we want
-    // to mutate and use `setState()` to update them.
-    // If we haven't created the note in our DB yet
-    // just remove it from the object locally
+  };
+
+  const addNote = () => {
+    setUuid(uuid - 1);
+    setNotesList([...notesList, { key: uuid }]);
+  };
+
+  const removeNote = (index: number) => {
+    const targetNote = notesList[index];
+    const { key } = targetNote;
+    const newNotes = [...notesList];
+    const newNotesObject = { ...notesObject };
+
     if (key < 0) {
-      delete notes[key];
+      delete newNotesObject[key];
     } else {
-      notes[key] = { isRemoved: true };
+      newNotesObject[key] = { isRemoved: true };
     }
 
-    this.setState(
-      {
-        notes,
-        existingNotes,
-      },
-      () => {
-        handleNotesChange(this.state);
-      }
-    );
-  }
-
-  renderNotes() {
-    const { existingNotes } = this.state;
-    return existingNotes.map((note, i) => (
+    newNotes.splice(index, 1);
```

Comment:
I notice that you're doing something different here than what the code used to do. In the previous code, it looked like we didn't change the length of `newNotes`/`notesList` (formerly named `existingNotes`), and instead we'd just mutate the item in place to set the `isRemoved` property to `true`.

I'm not sure what the consequences of changing the size of the array are, but if we wanted to play it safe, we should keep size of the `newNotes` array the same, and instead we should (immutably) change the object at that index. I think that would look like:

```js
newNotes[index] = {...targetNote, isRemoved: true};
```

Reply:
Yep, I changed it to delete the note and thus update the size of array on purpose. Previously, deleting a note created an annoying experience in the UI where the remaining note input labels would not update with the correct note index. This is because the label uses the index from the `existingNotes` array, which persisted the deleted note in its collection. See an screenshot below from our prod site. My update didn't seem to cause any downstream effects, fortunately.

<img width="990" alt="image" src="https://github.com/ShelterTechSF/askdarcel-web/assets/3012520/9138c79d-3a59-41be-b095-43a5a546a7b4">

Reply:
Ah, makes sense. Good to know that this fixed a latent bug!